Plandomizer Page

Veegie opened the pull request:
This PR adds a new page to the web app, providing a user-friendly means by which to create plando JSON files. I've attempted to match the look and feel of the seed generator, to make things slot in cleanly. (And also because it was much easier that way. 😛)

It's not at a level of polish where I'd call it "complete" yet, but I think it's functional and feature-complete enough for developer use, at the very least.

### Entry Point

![image](https://github.com/user-attachments/assets/e5781491-da14-49f0-80bd-7834c8f0ba2e)

A new link is added to the header bar for accessing the plandomizer.

The plandomizer interface is broken into four sections:

### Spirits & Chapters

![image](https://github.com/user-attachments/assets/136f5c81-9bcb-4c4b-b020-4292a2cd0e0e)

For setting required spirits, boss battles, difficulty scaling, and star power costs.

![image](https://github.com/user-attachments/assets/2b461f18-663c-4ab5-ae3f-faf4093a73dd)

These panels are a bit chunky, but these are "big" settings, so I think it works alright.

### Party

![image](https://github.com/user-attachments/assets/90b02956-7e0f-4c75-9076-82887a33b1e2)

For setting partner move costs. I waffled on making this a separate section, but it doesn't really fit anywhere else, especially with the Badges tab being pretty dense already.

### Badges

![image](https://github.com/user-attachments/assets/83e400ea-2584-4eb4-b595-36a695c05760)

For setting BP and FP costs for badges.

![image](https://github.com/user-attachments/assets/63b416ad-1029-4baa-8a47-c3fcdb4e1b59)

The columns of the layout will collapse on smaller screens.

### Items

![image](https://github.com/user-attachments/assets/de119298-b357-4706-9109-f7204c1aff8d)

For setting individual checks, items, and shop prices. Needless to say, this is a tough interface to make friendly, but I've done the best I can.

![image](https://github.com/user-attachments/assets/490acc9f-ac21-49ef-9cfb-2c32b84ddd11)

Shop items are displayed alongside a secondary input to specify their price.

![image](https://github.com/user-attachments/assets/58c937d6-3f41-4d91-b7a3-501628f30eff)

I added some filtering switches to help pare down the sheer number of inputs on the page, since not everyone who does a plando is interested in setting the contents of every bush and tree. As SuperBlocks and MultiCoin Blocks are not yet supported as per the plando validator readme, these checks are hidden by default.

![image](https://github.com/user-attachments/assets/5a07c5e2-db93-47e2-b055-e7df6e1ded77)

All of the item inputs use a material autocomplete (shamelessly pilfered from the starting items chiplist 😅) to make picking items a bit more painless.

### Generating

When finished choosing your plando settings, click the Download Plando File button to generate the JSON file and save it to your computer. No verbose validation currently takes place, but that can be added as a future enhancement. For now, the download button will simply be disabled if any input validation fails (e.g. FP cost greater than 75, BP cost greater than 10).

Output JSON is tidied to remove any null values or empty objects and arrays.

### Stray Implementation Notes

Organization within the components themselves is a bit messy. I wasn't sure how much to separate out larger static data like item and badge lists. I was hesitant to drop stuff into the entities folder if it might cause confusion or conflicts (e.g. my locations list & type vs the ItemLocation interface), but that's an easy fix if desired.

My locations list and the corresponding `CheckType` assignments are very much a best-guess and might require another pair of eyes to verify my categorizations. I'm not sure which specific yellow blocks are coin blocks, for example.

I've done my best to keep the layout and logic completely dependent on the data; if the underlying data changes, interface changes should be painless and automatic.

I fixed a bug with the `scrollToCenter` directive; previously, clicking inside the body of a tab group would cause the tab list to scroll unexpectedly. This can be observed on the seed generator site at present; make the browser window narrow so that the first settings card has a scrollbar for the tabs, swtich to the Comsetics tab, then click anywhere inside the cosmetics tab body. The scrolling action is now limited to when the user clicks on an actual tab button.

Any and all feedback is very welcome! I'm quite happy to be able to contribute to this project in some way.

### Next Steps

- [x] Make first-step validation a bit friendier and easier to grok.
- [ ] Add validation API call, and display validation warnings and messages alongside a download link.
- [ ] (Possibly?) Creation, saving, and loading of plando presets

Pronyo-Chan commented on app/src/app/pages/plando-page/plando-items/plando-items.component.ts:
```diff
@@ -0,0 +1,1218 @@
+      { name: "Hall to Ultra Boots (B3) - Yellow Block Left", type: CheckType.NORMAL },
+      { name: "Hall to Ultra Boots (B3) - Yellow Block Right", type: CheckType.NORMAL },
+      { name: "Ultra Boots Room (B3) - In Big Chest", type: CheckType.NORMAL },
+      { name: "Bridge to Shiver City (B2) - Yellow Block 1", type: CheckType.NORMAL },
```
These are all coin blocks. But also, aside from the first one, they are hidden blocks, idk how we should handle this. Is the Hidden block check type really necessary at all?

Pronyo-Chan commented on app/src/app/pages/plando-page/plando-items/plando-items.component.ts:
```diff
@@ -0,0 +1,1218 @@
+  {
+    name: "Koopa Region", checks: [
+      { name: "Pleasant Path Entry - Red Block Center", type: CheckType.NORMAL },
+      { name: "Pleasant Path Entry - Yellow Block Left", type: CheckType.NORMAL },
```
Coin block

Pronyo-Chan commented on app/src/app/pages/plando-page/plando-items/plando-items.component.ts:
```diff
@@ -0,0 +1,1218 @@
+
+  ngOnInit(): void {
+    this.initAutoComplete(this.itemsFormGroup);
+
```
Empty line can be removed

Pronyo-Chan commented on app/src/app/pages/plando-page/plando-items/plando-items.component.ts:
```diff
@@ -0,0 +1,1218 @@
+  public autocompleteSubscriptions: Subscription[] = [];
+  public searchText: FormControl;
+
+  ngOnInit(): void {
```
I like having explicits public for every function, although I might have forgotten in some parts of the codebase.
Looks like initAutoComplete can be private though

Pronyo-Chan replied:
(Same comments applies to other components where the visibility on functions is missing)

Pronyo-Chan commented on app/src/app/pages/plando-page/plando-page.component.scss:
```diff
@@ -0,0 +1,196 @@
+      }
+    }
+
+    &.tab-error {
```
This class doesn't seem to be actually used

Pronyo-Chan commented on app/src/app/pages/plando-page/plando-page.component.html:
```diff
@@ -0,0 +1,33 @@
+        </mat-card>
+        <mat-card class="settings-card">
+            <div class="button-container">
+                <button [disabled]="formGroup.invalid || isValidating" class="submit" type="button" (click)="onSubmit()">
```
A tooltip explaining there are errors when the button is disabled, like in the main rando page, would be nice

Pronyo-Chan commented on app/tsconfig.json:
```diff
@@ -14,6 +14,7 @@
     "module": "es2020",
     "lib": [
       "es2020",
+      "ES2021.String",
```
What did you need this for?

Veegie commented:
- Removed the Hidden Block check type (for some reason, I thought that was a rando option).
- Removed unused CSS rule.
- Fixed all Coin Block check types.
- Reverted `ES2021.String` lib import (added it for `String.replaceAll` but a global regex does the same thing).
- Added all missing access modifiers.
- Added a service for filtering inputs, along with a function for filtering numeric inputs. All numeric inputs with this filter will only accept inputs that result in a valid value, and will clamp to the input's min/max if they are exceeded. As such, it should no longer be possible to enter invalid values on any numeric inputs (Party tab, Badges tab, and shop item prices).
- Tidied up the autocomplete and made it slightly more performant.
- Added item name validation:
  ![image](https://github.com/user-attachments/assets/af7ba127-1c1c-423a-8ab9-6b6043a53095)
  - An invalid item name will highlight the input, its label, and all parent tab labels in red.
  - Valid names are any entries in the `PLANDO_ITEMS_LIST`, or `TRAP ([ValidExplicitItem])` where `[ValidExplicitItem]` is any entry in `PLANDO_ITEMS_LIST` except `TRAP`, `Consumable`, `NonProgression` or `Trash`.
- Added a tooltip on the Submit button to explain when validation errors are present.

Pronyo-Chan commented:
Would it be possible to only consider an item name invalid after the user is done editing? It seems a bit agressive to highlight it in red as soon as I type a character

Pronyo-Chan commented on app/src/app/pages/plando-page/plando-items/plando-items.component.html:
```diff
@@ -0,0 +1,60 @@
+                        <ng-container *ngIf="!filteredTypes.includes(check.type)">
+                            <div class="check-entry">
+                                <ng-container *ngIf="check.type !== CHECK_TYPES.SHOP">
+                                    <label>{{toDisplayString(check.name)}}: </label>
```
I just discovered a handy  tool in my bowser that allows throttling cpu to test performance, and on x4 slowdown, the app is quite unresponsive. I have a pretty good cpu so I think some actual users may experience the lag.

I narrowed it down to all the these pascalToVerbose calls made on the fly causing most of the lag. Every time a user action is taken, angular refreshed the whole view, including some components that are not even currently visible. This results in around 10k pascalToVerbose calls when a click on a text input in the items tab, for example. This action takes around a second to respond according to the performance monitoring, but by returning immediately from the pascalToVerbose function, it gets cut down to 500ms.

I suggest you should precompute all the toDisplayString calls in a dictionary for every label to improve performance. It could be done for the item names in the autocomplete too, but that seems responsible for only a small fraction of the calls. The labels are the issue as all of them get recomputed (probably multiple times) on each user action.

Veegie commented:
> The labels are the issue as all of them get recomputed (probably multiple times) on each user action.

Hoo boy. I noticed some slowdown when I was doing some profiling of my own, but I figured it was just the autocomplete and thus unavoidable. But the labels? Yeesh. I've memoized both `pascalToVerbose` and `toDisplayString`, and things seem to be much snappier (though the page still seems a bit sluggish when I throttle my own CPU 4x, admittedly).

> Would it be possible to only consider an item name invalid after the user is done editing? It seems a bit agressive to highlight it in red as soon as I type a character

I had the same feeling, and after a couple of hours of effort, I think 4eb5275 is probably the best I can do; name inputs will no longer validate immediately the first time you start typing into them, but will do so on subsequent edits (even if you just focus, blur, and then re-focus, unfortunately). In addition, there's a flash of red when selecting an autocomplete option. I've tried a number of other options, none of which panned out any better:
- Restricting the validator to only firing on `blur` events works if you manually type in an item name, but selecting an autocomplete option will result in the validation being done on the value prior to the one the autocomplete sets. The field will be valid but display as invalid until the input is focused and blurred again.
- With the above approach, I attempted to suppress `blur` events when the autocomplete panel is open, but that doesn't seem to be possible, and it wouldn't have fixed the need to refocus and reblur to actually get the validation to go through anyway.
- Just restricting the validator to only firing on `touched` form controls prevents the aggressive initial validation, but also prevents validation on the first value entered entirely.

It's kind of a shame that the material autocomplete doesn't play nice with validators, but c'est la vie, I suppose.

Veegie commented:
Awesome, thanks! I'll get started on saving and loading next and do it as a separate PR.
